cmd/thanos/main.go

```diff
@@ -161,9 +169,79 @@ func joinCluster(logger log.Logger, typ cluster.PeerType, bindAddr, advertiseAdd
 	// TODO(fabxc): generate human-readable but random names?
 	name, err := ulid.New(ulid.Now(), rand.New(rand.NewSource(time.Now().UnixNano())))
 	if err != nil {
-		return nil, err
+		return cluster.JoinConfig{}, err
 	}
-	return cluster.NewPeer(logger, name.String(), typ, bindAddr, advertiseAddr, apiAddr, peers)
+	return cluster.JoinConfig{
+		Name:          name.String(),
+		Addr:          bindAddr,
+		AdvertiseAddr: advertiseAddr,
+		KnownPeers:    resolvedPeers,
+	}, nil
+}
+
+func resolvePeers(ctx context.Context, peers []string, myAddress string, res net.Resolver, waitIfEmpty bool) ([]string, error) {
+	var resolvedPeers []string
+
+	for _, peer := range peers {
+		host, port, err := net.SplitHostPort(peer)
+		if err != nil {
+			return nil, errors.Wrapf(err, "split host/port for peer %s", peer)
+		}
+
+		retryCtx, cancel := context.WithCancel(ctx)
+		if ips, err := res.LookupIPAddr(ctx, host); err == nil {
+			if len(ips) == 0 {
+				var lookupErrSpotted bool
+
+				err := runutil.Retry(2*time.Second, retryCtx.Done(), func() error {
+					if lookupErrSpotted {
+						cancel()
+					}
+
+					ips, err = res.LookupIPAddr(retryCtx, host)
+					if err != nil {
+						lookupErrSpotted = true
+						return errors.Wrapf(err, "IP Addr lookup for peer %s", peer)
+					}
+
+					ips = removeMyAddr(ips, port, myAddress)
+					if len(ips) == 0 {
+						if !waitIfEmpty {
+							return nil
+						}
+						return errors.New("empty IPAddr result. Retrying")
+					}
+
+					return nil
+				})
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			for _, ip := range ips {
+				resolvedPeers = append(resolvedPeers, net.JoinHostPort(ip.String(), port))
+			}
+		}
+
+		// Assume direct address.
+		resolvedPeers = append(resolvedPeers, peer)
+	}
+
+	return resolvedPeers, nil
+}
```

Comment:
Quite a few indentions. I think we can reduce that by a few by swapping the conditions and using `continue` etc.